### PR TITLE
Adds full colour 80 column support.

### PR DESCRIPTION
**Pull request by thejpster**
Currently the OS doesn't know about it so that comes up grey-on-brown. But the boot screen looks lovely.

image

### CODE REVIEW

**Comment by thejpster** [comment]
image

**Comment by thejpster** [comment]
Reviews welcome, in case I missed a comment about the clock frequency or got something wrong.

File: src/vga/mod.rs [hunk 1]
```diff
@@ -271,6 +272,19 @@ static CORE1_ENTRY_FUNCTION: [u16; 2] = [
 	0x46c0, // nop - pad this out to 32-bits long
 ];
 
+/// Holds the colour look-up table for text mode.
+///
+/// The input is a 9-bit vlaue comprised of the 4-bit foreground colour index,
```
Comment (thejpster) [inline src/vga/mod.rs:277]:
```suggestion
/// The input is a 9-bit value comprised of the 4-bit foreground colour index,
```

**Review by Detegr — APPROVED** [review]
Looks good to me!

File: Cargo.toml [hunk 2]
```diff
@@ -55,6 +55,8 @@ codegen-units = 1
 debug = true
 # better optimizations
 lto = true
+# set the optimisation level
+opt-level = 3
```
[the remark this answers is not in the document]
Comment (thejpster) [reply Cargo.toml:59]: I think it's the default, but it probably won't work at -O2 so we should specify it. It definitely doesn't work at -Os

[the remark this answers is not in the document]
Comment (jannic) [reply Cargo.toml:59]: > It definitely doesn't work at -Os

Thats very interesting indeed, as in my impression code for the rp2040 often works as fast, if not faster, with -Os.
This is just anecdotal, though, I didn't do any systematic measurements, but only compared a few cases.
Do you know which code doesn't get optimized well enough with -Os?

[the remark this answers is not in the document]
Comment (thejpster) [reply Cargo.toml:59]: My only debug was looking at the pulse on the LED and seeing it was no longer shorter than the length of a scan line. But I will go back and get the numbers for each opt-level for the record.

[the remark this answers is not in the document]
Comment (thejpster) [reply Cargo.toml:59]: | Opt Level | Draw Time | Percent | Binary Size (inc OS) |
|-----------|-----------|---------|----------------------|
| 3         | 28.4µs    | 89.2%   | 112 KiB              |
| 2         | 28.1µs    | 88.6%   | 108 KiB              |
| 1         | 33.8µs    | > 100%  | 116 KiB              |
| s         | 28.3µs    | 89.3%   | 104 KiB              |

`opt-level = "s"` makes the most sense then.

**Comment by jonathanpallant** [comment]
The stability problems I have (#49) occur on develop, so no reason to hold off on this PR.

**Pull request merged**